### PR TITLE
drivers: watchdog: Adding watchdog support for sam SOC

### PR DESCRIPTION
**Pull request by subbu147**
Adding watchdog driver for sam SoC. The current implementation just
disables the watchdog on device boot.
This PR adds the following support for SAM3X, SAM4S and SAME70
1. Activate processor reset
2. Activate all reset
3. Generate interrupt
on watchdog timer expiration.

Signed-off-by: Subramanian Meenakshi Sundaram <subbu147@gmail.com>

### CODE REVIEW

**Review by dbkinder — APPROVED** [review]

**Review by dbkinder — APPROVED** [review]
+1 for doc changes

File: drivers/watchdog/wdt_sam.c [hunk 1]
```diff
@@ -22,25 +24,79 @@
...
+	if ((timeout < min) || (timeout > max)) {
+		LOG_ERR("Invalid timeout value allowed range:"
+			"%d ms to %d ms", min / 1000, max / 1000);
+		return -EINVAL;
+	}
+
+	return WDT_MR_WDV(timeout / min);
+}
+
 static void wdt_sam_enable(struct device *dev)
```
Comment (galak) [inline drivers/watchdog/wdt_sam.c]: Just remove wdt_sam_enable()

**Comment by galak** [comment]
I pushed some updates and reworked the driver a bit, take a look and test it out if you can.

**Review by nashif — DISMISSED** [review]

File: samples/drivers/watchdog/prj.conf [hunk 2]
```diff
@@ -1 +1,2 @@
 CONFIG_WATCHDOG=y
+CONFIG_WDT_SAM_DISABLE_AT_BOOT=n
```
[the remark this answers is not in the document]
Comment (subbu147) [reply samples/drivers/watchdog/prj.conf]: Removed the vendor specific configuration from the sample config

**Comment by subbu147** [comment]
@galak Your changes holds good. I have tested All the modes and all the new negative(fault handling) scenarios.

**Review by galak — APPROVED** [review]

File: drivers/watchdog/wdt_sam.c [hunk 3]
```diff
@@ -22,76 +24,231 @@
...
+	 * Once disabled, it cannot be enabled until the device is reset
+	 */
 	wdt->WDT_MR |= WDT_MR_WDDIS;
+	data->mode_set = true;
 
 	return 0;
 }
 
-static int wdt_sam_set_config(struct device *dev, struct wdt_config *config)
+int wdt_sam_setup(struct device *dev, u8_t options)
```
Comment (MaureenHelm) [inline drivers/watchdog/wdt_sam.c]: This should be static

File: drivers/watchdog/wdt_sam.c [hunk 4]
```diff
@@ -22,76 +24,231 @@
...
+	}
+
+	wdt->WDT_MR = data->mode;
+	data->mode_set = true;
+
+	return 0;
 }
 
-static void wdt_sam_get_config(struct device *dev, struct wdt_config *config)
+int wdt_sam_install_timeout(struct device *dev,
```
Comment (MaureenHelm) [inline drivers/watchdog/wdt_sam.c]: This should be static

File: drivers/watchdog/wdt_sam.c [hunk 5]
```diff
@@ -22,76 +24,231 @@
...
+		     WDT_MR_WDV(timeout_value) |
+		     WDT_MR_WDD(timeout_value);
+
+	data->timeout_valid = true;
+
+	return 0;
 }
 
-static void wdt_sam_reload(struct device *dev)
+int wdt_sam_feed(struct device *dev, int channel_id)
```
Comment (MaureenHelm) [inline drivers/watchdog/wdt_sam.c]: Static

File: drivers/watchdog/wdt_sam.c [hunk 6]
```diff
@@ -22,76 +24,231 @@
...
+	.setup = wdt_sam_setup,
 	.disable = wdt_sam_disable,
-	.get_config = wdt_sam_get_config,
-	.set_config = wdt_sam_set_config,
-	.reload = wdt_sam_reload
+	.install_timeout = wdt_sam_install_timeout,
+	.feed = wdt_sam_feed,
+};
+
+struct wdt_sam_dev_cfg wdt_sam_cfg = {
```
Comment (MaureenHelm) [inline drivers/watchdog/wdt_sam.c]: This should be static const

File: dts/bindings/watchdog/atmel,sam-watchdog.yaml [hunk 7]
```diff
@@ -0,0 +1,39 @@
+---
```
Comment (MaureenHelm) [inline dts/bindings/watchdog/atmel,sam-watchdog.yaml]: Add a license/copyright header

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/watchdog/wdt_sam.c (hunk 4 above)
Comment (subbu147) [reply drivers/watchdog/wdt_sam.c]: addressed

File: drivers/watchdog/wdt_sam.c (hunk 5 above)
Comment (subbu147) [reply drivers/watchdog/wdt_sam.c]: addressed

File: drivers/watchdog/wdt_sam.c (hunk 6 above)
Comment (subbu147) [reply drivers/watchdog/wdt_sam.c]: addressed

File: dts/bindings/watchdog/atmel,sam-watchdog.yaml (hunk 7 above)
Comment (subbu147) [reply dts/bindings/watchdog/atmel,sam-watchdog.yaml]: addressed

**Comment by galak** [comment]
@MaureenHelm can you re-review

File: drivers/watchdog/wdt_sam.c [hunk 8]
```diff
@@ -22,76 +24,231 @@
...
+			LOG_ERR("Invalid(NULL) ISR callback passed\n");
+			return -EINVAL;
+		}
+		break;
+
+		/* Processor only reset mode not available in same70 series */
+#ifdef WDT_MR_WDRPROC
+	case WDT_FLAG_RESET_CPU_CORE:
+		/*A Watchdog fault activates the processor reset*/
+		printk("Configuring reset CPU only mode\n");
```
Comment (MaureenHelm) [inline drivers/watchdog/wdt_sam.c]: This should be LOG_DBG

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
One small change needed, but otherwise looks good.

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**